### ISSUE TITLE
Issue with ES Hive exporter in the case of Parent-Child Index

### ISSUE THREAD

**Issue by dkjhanitt** [elastic/elasticsearch-hadoop#575]
I have an index with 2 different document types A and B where B is child of  A. I am trying to use elasticsearch-hadoop jar to index the data into ES from hive table. Here is a mapping for B

``` javascript
"B": {
            "_parent": {"type": "A"},
            "properties": {
                  "type_1": {"properties": {
                        "field1": {"type": "integer" },
                        "field2": { "type": "integer" }
                    }
                }
        }
```

I have a hive table with 2 columns storing its parent and a JSON containing the list of key value pairs 

```
hive > select * from MY_TABLE;
parent_id   {"type_1": [{"field1":341 ,"field2":1404},{"field1":156,"field2":1403},{"field1":1350,"field2":1401}]}      
As you can see the JSON is legit here.
```

Here is my hive query which is trying to index data into ES

```
CREATE EXTERNAL TABLE MY_TABLE_es
(
    parent     BIGINT,
    xyz    STRING
)
ROW FORMAT SERDE 'org.elasticsearch.hadoop.hive.EsSerDe'
STORED BY 'org.elasticsearch.hadoop.hive.EsStorageHandler'
TBLPROPERTIES('es.nodes' = 'localhost',
        'es.port' = '9200',
         'es.resource.write' = 'my_index/B',
         'es.input.json'='yes',
        'es.mapping.routing'='parent',
        'es.mapping.id' = 'parent');

INSERT overwrite table MY_TABLE_es select parent, xyz  from MY_TABLE ;      
```
## Error that I am getting

```
Diagnostic Messages for this Task:
java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: Hive Runtime Error while processing row {"parent":421,"type_1":"{\"type_1\":[{\"field1\":341,\"field2\":1404},{\"field1\":156,\"field2\":1403},{\"field1\":1350,\"field2\":1401}]}"}
    at org.apache.hadoop.hive.ql.exec.ExecMapper.map(ExecMapper.java:159)
    at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:50)
    at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:417)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:332)
    at org.apache.hadoop.mapred.Child$4.run(Child.java:268)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:396)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1408)
    at org.apache.hadoop.mapred.Child.main(Child.java:262)
Caused by: org.apache.hadoop.hive.ql.metadata.HiveException: Hive Runtime Error while processing row {"parent":8577236,"type_1":"{\"type_1\":[{\"field1

FAILED: Execution Error, return code 2 from org.apache.hadoop.hive.ql.exec.MapRedTask
MapReduce Jobs Launched:
Job 0: Map: 1   HDFS Read: 0 HDFS Write: 0 FAIL
Total MapReduce CPU Time Spent: 0 msec


```

I am generating json in php (through streaming) using json_encode($array) and my JSON looks legit. Still not sure why am i facing this issue. Can anyone please help.

**Comment by costin** [comment]
Yeah, the error is not really helpful - you'll find more information in the Hive logs (see the files that are created by the job).
Likely the issue is caused by the use of `es.mapping.id` - the value is extracted from the payload (in this case the json document) which should contain a token called `parent`. But I don't see one.

Note that you don't need to create the JSON yourself - you can send the data as is as a CSV which Hive can easily load as columns and then pass that over to the connector which creates the json as needed.
The advantage is not only you don't have to create the JSON (we do it for you and likely more efficient) but also there's a lot more freedom in creating the final response that is sent out to ES.
Basically if you use JSON input, the connector is not going to modify it in any way - however if its the connector that creates the JSON instead, it can do obviously structure it accordingly.

**Comment by dkjhanitt** [comment]
Hi Costin,
Thanks for the response. So let me explain you in detail 
I have an index which has 2 document types i) country ii) state, where state is child of the country.
Each state has an object called city which is nothing but list of (city_name: String, zipcode: Integer ). In terms of ES mapping state is represented as 
```javascript
"state": {
            "_parent": {"type": "country"},
            "properties": {
                  "city": {
                   "properties": {
                        "city_name": {"type": "string", "index": "not_analyzed" },
                        "zipcode": { "type": "integer" }
                    }
                }
        }

```

## How data is present in the hive table
One of the column in hive table is country and another one is colon separated city_name and zipcode which is comma-separated with other city_name and zipcode. So it looks like
SanFrancisco:94107,RedwoodCity:94063,SanMateo:94401

## How my hive external table looks like
```javascript
{
"city": [
             { "city_name": "SanFrancisco", "zipcode": 94107},
             { "city_name": "RedwoodCity", "zipcode": 94063},
             { "city_name": "SanMateo", "zipcode": 94401},
          ], 
  "parent": "USA"
}
 ```
```
DROP table if EXISTS ${hiveconf:COUNTRY_TABLE}_es;
CREATE EXTERNAL TABLE ${hiveconf:COUNTRY_TABLE}_es
(
    country     string,
    cities         string
)
ROW FORMAT SERDE 'org.elasticsearch.hadoop.hive.EsSerDe'
STORED BY 'org.elasticsearch.hadoop.hive.EsStorageHandler'
TBLPROPERTIES('es.nodes' = 'localhost',
        'es.port' = '9200',
         'es.resource.write' = 'my_index/state',
         'es.input.json'='yes',
        'es.mapping.routing'='country',
        'es.mapping.id' = 'country');

INSERT overwrite table ${hiveconf:COUNTRY_TABLE}_es select country, cities  from ${hiveconf:COUNTRY_TABLE} ;
```

I believe this is going to give you more clarity on what is the correct way to do it. We are using hive 0.10. Can you please help me out in fixing this issue.

**Comment by costin** [comment]
@dkjhanitt Unfortunately without knowing the actual Hive error, I am in the dark. In particular since with each release the semantics of Hive internals tend to change (by the way, the latest version is 1.2.1 and quite a few things changed since 0.10).

Regarding es-hadoop you can use it to add data in ES with any of its integrations - MR, Cascading, Hive, Pig, etc...
The bulk format (as [described](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) by the ES documentation) is not supported since its the connector that has to generate the metadata.
However if your documents are already in JSON format then that is supported (it might be supported by the ES clients but in that case there won't be any parallel insert).

Do note that if Hive is problematic and you don't use it table for writing, you can try a different approach such as using pure M/R (it's not that big of a deal especially if the data is already parsed) or even Pig.
There's also Spark but I don't want to derail you.

But before pivoting, try to get a hold of the actual exception that occurs on the Hive server - that will clarify a LOT of things.

Cheers,

**Comment by dkjhanitt** [comment]
Hi Costin,
I was able to fix the issue. Iused brickhouse jar to create an Array<struct<country:string, city:string>> to fix the issue.
Just checking with you if there is a way to delete the document using elasticsearch-hive ? 

Thanks,
Deepak

**Comment by costin** [comment]
> I was able to fix the issue. Iused brickhouse jar to create an Array> to fix the issue.

Can you offer more details on what was the solution?

> Just checking with you if there is a way to delete the document using elasticsearch-hive ?

Delete operation is not currently supported by the connector.

**Comment by dkjhanitt** [comment]
So I have an intermediate hive table with columns 
country, city_name, zipcode
```
create temporary function collect as 'brickhouse.udf.collect.CollectUDAF';

CREATE TABLE my_new_table
  AS
SELECT
      country as country,
       collect(named_struct("city_name", cityname, "zipcode", zipcode)) as cities_with_zipcode
FROM 
      my_table GROUP BY country;
```
So now my_new_table had 2 columns. I can use it to feed my hive external es table using insert query.
```
country string,
cities_with_zipcode array<struct<city_name: string, zipcode: bigint>>
```
There was another mistake, since my index is structured as parent-child, I had to include following details while creating an external table
```
'es.mapping.id' = 'country',
'es.mapping.parent' = 'country',
```
IIRC the hive error I was getting did not mention about missing parent details. 

Do you recommend a way to delete documents from ES ? As of now I have an ETL framework (in scala) based on akka actor which uses ES REST endpoint to delete the document. I was hoping that elasticsearch-hadoop jar has some way to delete the documents. Please suggest if there is anyway I can do it. 

Hope it clarifies.
Thanks,
Deepak

**Comment by costin** [comment]
> IIRC the hive error I was getting did not mention about missing parent details. 

As I've mentioned twice in the responses above, the Hive logs indicate the proper not the error (which in fact doesn't provide any relevant information, such as the fact that you are using ES as an external table).

**Comment by jbaiera** [comment]
I'm going to go ahead and close this out since it seems that the conversation has trailed off. If this is still a problem, the issue can be re-opened.
